### PR TITLE
 Fix the plugin command structure (single-level, not nested)

### PR DESCRIPTION
**Pull request by andreas-kupries**
Ref: https://trello.com/c/FQ3KvVZ1/619-usb-plugin-cli-command-set-fixes

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -36,6 +36,13 @@ func main() {
 func (c *UsbPlugin) Run(cliConnection plugin.CliConnection, args []string) {
 	c.argLength = len(args)
 
+	if c.argLength < 2 {
+		if c.argLength > 0 && strings.HasPrefix(args[0], "CLI-MESSAGE-") {
+			// Internal CLI command (e.g. uninstall); don't show help text
+		}
+		return
+	}
```
Comment (mook-as) [inline main.go]: This chunk basically means any command without arguments is just busted.

File: main.go [hunk 2]
```diff
@@ -216,17 +207,17 @@ OPTIONS:
...
+					Usage: "cf usb-delete-driver-endpoint NAME",
 				},
 			},
 			plugin.Command{
-				Name:     "usb update-driver-endpoint",
+				Name:     "usb-update-driver-endpoint",
 				HelpText: "Update a driver instance",
 				UsageDetails: plugin.Usage{
-					Usage: `cf usb update-driver-endpoint NAME [-t ENDPOINT_URL] [-k AUTHENTICATION_KEY] [-c METADATA_AS_JSON]
+					Usage: `cf usb-update-driver-endpoint NAME [-t ENDPOINT_URL] [-k AUTHENTICATION_KEY] [-c METADATA_AS_JSON]
```
Comment (mook-as) [inline main.go:220]: The help text below still says `cf usb update-driver-endpoint…`; see also the other help texts.

**Review by mook-as — CHANGES_REQUESTED** [review]
I don't know how this command works, but `cf usb-create-driver-endpoint foo http://foo.cf-dev.io/ key` died on me with a traceback.

File: main.go (hunk 2 above)
Comment (andreas-kupries) [reply main.go:220]: Will fix.

File: main.go (hunk 1 above)
Comment (andreas-kupries) [reply main.go]: Will check. Guess the condition has to check the command as well.
I better check the backup/restore plugin also.

File: main.go (hunk 2 above)
Comment (andreas-kupries) [reply main.go:220]: Done with 563bb1b72c3fa3ec8f806e6e4937f56b528ab47a

File: main.go (hunk 1 above)
Comment (andreas-kupries) [reply main.go]: Backup/restore looks good. Decided to snarf its code for here.
Done with commit d7c7b547d64d1b1bb5bd5302b23a4dc6c9bc577c

**Review by mook-as — CHANGES_REQUESTED** [review]
```
$ ./build/linux-amd64/cf-usb-plugin 
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/SUSE/cf-usb-plugin/vendor/github.com/cloudfoundry/cli/plugin.Start(0x930300, 0xc420426720)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/vendor/github.com/cloudfoundry/cli/plugin/plugin_shim.go:16 +0x39c
main.main()
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/main.go:32 +0x3d
```

**Comment by andreas-kupries** [comment]
Ok. Direct execution of the plugin outside of the cf context is not really supported.
Still, not a good idea to simply crash either, so I will add code to detect and handle that.
With a nice error message.

As for `but cf usb-create-driver-endpoint foo http://foo.cf-dev.io/ key died on me with a traceback.`, do you still have that traceback ?

Note that I do not really know how the commands work either, and I have not changed anything below the cobra/viper setup, i.e. the command dispatch itself.

**Comment by mook-as** [comment]
Traceback:
```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/SUSE/cf-usb-plugin/cmd.glob..func1(0xb6e580, 0xc4202301e0, 0x3, 0x3)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/cmd/create-driver-endpoint.go:29 +0x56f
github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra.(*Command).execute(0xb6e580, 0xc4202300f0, 0x3, 0x3, 0xb6e580, 0xc4202300f0)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra/command.go:565 +0x20d
github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xb6ed80, 0x4, 0xc420432380, 0x8d70c6)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra/command.go:651 +0x2d4
github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra.(*Command).Execute(0xb6ed80, 0x4, 0x6)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/vendor/github.com/spf13/cobra/command.go:610 +0x2b
github.com/SUSE/cf-usb-plugin/cmd.Execute()
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/cmd/root.go:29 +0x2d
main.(*UsbPlugin).Run(0xc42051f0e0, 0x936d00, 0xc4203b48e0, 0xc42001e140, 0x4, 0x4)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/main.go:125 +0x49d
github.com/SUSE/cf-usb-plugin/vendor/github.com/cloudfoundry/cli/plugin.Start(0x930300, 0xc42051f0e0)
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/vendor/github.com/cloudfoundry/cli/plugin/plugin_shim.go:30 +0x28f
main.main()
        /home/marky/src/work/src/github.com/SUSE/cf-usb-plugin/main.go:32 +0x3d
```

Not sure we should fix this in this PR.

**Comment by andreas-kupries** [comment]
Regarding direct execution, the backup plugin is fine, and has a nice error message.
A message which comes out of the plugin shim.

```
andreask@ten:/fast/go/src/github.com/SUSE/cf-plugin-backup> ./build/linux-amd64/cf-plugin-backup 
This cf CLI plugin is not intended to be run on its own

andreask@ten:/fast/go/src/github.com/SUSE/cf-plugin-backup> grep -rn 'This cf CLI plugin is not intended to be run on its own' .
./vendor/code.cloudfoundry.org/cli/plugin/plugin_shim_test.go:27:			Expect(session).To(gbytes.Say("This cf CLI plugin is not intended to be run on its own"))
./vendor/code.cloudfoundry.org/cli/plugin/plugin_shim.go:17:		fmt.Printf("This cf CLI plugin is not intended to be run on its own\n\n")
```

Have to see if the usb plugin does something different. Or maybe an older shim ?

Regarding the issue with the driver endpoint I am very much ok with having this out of scope for this PR and ticket.

**Comment by andreas-kupries** [comment]
Right the usb plugin does not have the shim vendored in.
And I have not updated whatever plugin stuff I have in my `GOPATH` in forever.
This needs some more investigation if the usb plugin is actually uptodate with all the cli integration.

**Comment by andreas-kupries** [comment]
The usb plugin uses the old `v6.17.1-50-gcc19f70` of `github.com/cloudfoundry/cli`.
The backup plugin uses `v6.35.0` of `code.cloudfoundry.org/cli`, which apparently is an alias or some such of  `github.com/cloudfoundry/cli`. So, still the same repo underneath, different versions.

usb plugin uses Godeps.
backup plugin uses `dep` ? (`go dep` ?). (Files `Gopkg.{toml,lock}`).

**Comment by andreas-kupries** [comment]
We decided to put the change `Godeps` 2 `dep`, and plugin shim update into a separate ticket.
That means this is ok as is (no worse than before).

**Review by mook-as — APPROVED** [review]
Bumping the direct execution out of this PR, as that requires upgrading the dependencies. We can do that in a separate PR.

**Pull request merged**